lib/modules/datasource/packagist/index.ts

```diff
@@ -119,27 +118,29 @@ export class PackagistDatasource extends Datasource {
     return packagistFile;
   }
 
-  private static extractDepReleases(versions: RegistryFile): ReleaseResult {
-    const dep: ReleaseResult = { releases: [] };
-    // istanbul ignore if
-    if (!versions) {
-      return dep;
+  /* istanbul ignore next */
+  private static extractDepReleases(
+    composerReleases: unknown
+  ): ReleaseResult | null {
+    const parsedRecord =
+      schema.ComposerReleasesRecord.safeParse(composerReleases);
+    if (parsedRecord.success) {
+      logger.trace('Packagist: extracting releases from record');
+      return schema.extractReleaseResult(Object.values(parsedRecord.data));
     }
-    dep.releases = Object.keys(versions).map((version) => {
-      // TODO: fix function parameter type: `versions`
-      const release = (versions as any)[version];
-      const parsedVersion = release.version ?? version;
-      dep.homepage = release.homepage || dep.homepage;
-      if (release.source?.url) {
-        dep.sourceUrl = release.source.url;
-      }
-      return {
-        version: parsedVersion.replace(regEx(/^v/), ''),
-        gitRef: parsedVersion,
-        releaseTimestamp: release.time,
-      };
-    });
-    return dep;
+
+    const parsedArray =
+      schema.ComposerReleasesArray.safeParse(composerReleases);
+    if (parsedArray.success) {
+      logger.trace('Packagist: extracting releases from array');
+      return schema.extractReleaseResult(parsedArray.data);
+    }
+
+    logger.trace(
+      { composerReleases },
+      'Packagist: unknown format to extract from'
```

Comment:
In this case we can assume we will essentially never notice it unless specifically looking for it with the trace turned on. Is that ok?

Reply:
Can we use `trace` level to oserve whether we hit this branch in real or not? I thought to remove all unused branches in this function in 3 days or so.

Reply:
We definitely cannot log at trace level in the app - it's to big. We do this:
- Log at debug level, save to file (S3), available on demand via the web UI
- Send all info and above logs to cloudwatch, which we can filter on

If you log this at debug level, and I rewrite the log level to info, we could filter based on it in the production app. But I'd want you to try to make sure that it only logs the error once per repo and not e.g. 100s or 1000s of times.

Reply:
Okay, so I think I'll create spin-off PR for utility that helps to log something once per repo. Seems like it can be helpful in similar situations outside of packagist case.

Reply:
"Log once per repository per run" sounds like an easier one. Gets harder if you want one log message but aggregated & deduplicated metadata and/or a different uniqueness key (e.g. once per repo + package file). But such a feature would be good for everyone's experience

Reply:
So, what we want here:
- For the execution branch that we can hit locally with debugger, no logging
- For the execution branches we cannot debug/observe, log at the `info` level once per repo

After couple days of observations, we remove branches that weren't logged and retain the executed ones, removing logging